rust/datafusion/src/execution/physical_plan/expressions.rs

```diff
@@ -991,48 +991,36 @@ impl fmt::Display for BinaryExpr {
     }
 }
 
-// Returns a formatted error about being impossible to coerce types for the binary operator.
-fn coercion_error<T>(
-    lhs_type: &DataType,
-    op: &Operator,
-    rhs_type: &DataType,
-) -> Result<T> {
+// Returns a formatted error about being impossible to coerce types to a common type
+fn coercion_error<T>(lhs_type: &DataType, rhs_type: &DataType) -> Result<T> {
     Err(ExecutionError::General(
         format!(
-            "The binary operator '{}' can't evaluate with lhs = '{:?}' and rhs = '{:?}'",
```

Comment:
The message was re-formulated because it made it a bit clearer that the coercion did not fail due to the use of the types on a specific operator: any operator would have yielded the same error, and thus I dropped the operator from the message altogether.

This was a side effect of removing the operator from the coercions' functuons. Essentially, code such as `numerical_coercion(lhs_type: &DataType, op: Operator, rhs_type: &DataType)` suggests that coercion depends on the operator, but this is not true: only the error message depends on the operator.

However, I do agree with you that that is a generic message. We could wrap the error and re-write it with the operator's information. E.g.

```
Operator::Plus | Operator::Minus | Operator::Divide | Operator::Multiply => {
            numerical_coercion(lhs_type, rhs_type)
        }
```

could be a place to catch the error and re-write the error message with the operator's information. What do you think?

Reply:
I guess I was thinking about the error from the user's perspective rather than the implementation. If I were a user and I got this error message, I want to know what part of my query / expression to fix.

For example, if I had `(A + B) - (C * D)` and I got "can't coerce int to float" error I am not sure I would know if the problem was with `A + B`or with `C * D` or with the overall difference. While the ideal error message would perhaps include an offset and a partial output of the original string, but having the operator name would make it easier to track down what I needed to do to fix my query

Reply:
I don't have any opinion on how the context of the error gets into the error message, but it does seem important to me

Reply:
I agree that this seems like a regression in UX. It is useful to know which operator was involved to reduce the effort to debug an error like this.

Reply:
I also agree. I've added a test to the error message to contain the operator, and made code pass the test.